sway/input/keyboard.c

```diff
@@ -204,71 +230,64 @@ static void handle_keyboard_key(struct wl_listener *listener, void *data) {
-			config->current_mode->keycode_bindings,
-			code_modifiers, input_inhibited);
-	struct sway_binding *translated_binding = get_active_binding(
-			&keyboard->state_keysyms_translated,
-			config->current_mode->keysym_bindings,
-			translated_modifiers, input_inhibited);
-	if (translated_binding && !binding) {
-		binding = translated_binding;
-	} else if (binding && translated_binding && binding != translated_binding) {
-		wlr_log(L_DEBUG, "encountered duplicate bindings %d and %d",
-			binding->order, translated_binding->order);
-	}
-	struct sway_binding *raw_binding = get_active_binding(
-			&keyboard->state_keysyms_raw,
-			config->current_mode->keysym_bindings,
-			raw_modifiers, input_inhibited);
-	if (raw_binding && !binding) {
-		binding = raw_binding;
-	} else if (binding && raw_binding && binding != raw_binding) {
-		wlr_log(L_DEBUG, "encountered duplicate bindings %d and %d",
-			binding->order, raw_binding->order);
-	}
+	struct sway_binding *binding_pressed = NULL;
+	get_active_binding(&keyboard->state_keycodes,
+			config->current_mode->keycode_bindings, &binding_pressed,
+			code_modifiers, false, input_inhibited);
+	get_active_binding(&keyboard->state_keysyms_translated,
+			config->current_mode->keysym_bindings, &binding_pressed,
+			translated_modifiers, false, input_inhibited);
+	get_active_binding(&keyboard->state_keysyms_raw,
+			config->current_mode->keysym_bindings, &binding_pressed,
+			raw_modifiers, false, input_inhibited);
+
+	struct sway_binding *binding_released = NULL;
+	get_active_binding(&keyboard->state_keycodes,
+			config->current_mode->keycode_bindings, &binding_released,
+			code_modifiers, true, input_inhibited);
+	get_active_binding(&keyboard->state_keysyms_translated,
+			config->current_mode->keysym_bindings, &binding_released,
+			translated_modifiers, true, input_inhibited);
+	get_active_binding(&keyboard->state_keysyms_raw,
+			config->current_mode->keysym_bindings, &binding_released,
+			raw_modifiers, true, input_inhibited);
 
 	bool handled = false;
 
-	// Execute the identified binding if need be.
-	if (keyboard->held_binding && binding != keyboard->held_binding &&
+	// Execute the stored binding, clear stored binding, and store new binding
+	if (keyboard->held_binding && binding_released != keyboard->held_binding &&
 			event->state == WLR_KEY_RELEASED) {
 		keyboard_execute_command(keyboard, keyboard->held_binding);
 		handled = true;
 	}
-	if (binding != keyboard->held_binding) {
+	if (binding_released != keyboard->held_binding) {
 		keyboard->held_binding = NULL;
 	}
-	if (binding && event->state == WLR_KEY_PRESSED) {
-		if (binding->release) {
-			keyboard->held_binding = binding;
-		} else {
-			keyboard_execute_command(keyboard, binding);
-			handled = true;
-		}
+	if (binding_released && event->state == WLR_KEY_PRESSED) {
+		keyboard->held_binding = binding_released;
+	}
+
+	// Execute the current pressed binding
+	if (binding_pressed && event->state == WLR_KEY_PRESSED) {
```

Comment:
It seems that we don't care about `binding_pressed` when `event->state != WLR_KEY_PRESSED`. Maybe we can move the code computing it in this `if`?